### PR TITLE
JS: add dataflow barrier for if(!x)

### PR DESCRIPTION
**Pull request by erik-krogh**
Consider the example: 

```JavaScript
var source = ...
if (source) {
  source = sanitize(source)
}
sink(source);
```

Previously we would detect flow from `source` to `sink`, as it can pass unsanitized through the "else" branch of the if. 

Looking at how to solve this I encountered a barrier that @asgerf wrote in `PrototypePollutionUtility.ql` that solved the exact issue.  

I moved that barrier into an `AdditionalBarrierGuardNode` that works for all DataFlow configurations.  
(It looks very different when moved from `isBarrier` to a `AdditionalBarrierGuardNode`, but it does the same thing).

### CODE REVIEW

**Comment by erik-krogh** [comment]
Given how fundamental the change is, we will need a thorough evaluation.  

But I would prefer some feedback before i start that evaluation.

**Comment by asgerf** [comment]
I agree that this should be the default behavior for taint-tracking configurations 👍 

For general data-flow configurations I'm not so sure.
Unfortunately there's no way to opt out of a standard barrier guard, so this means it's no longer possible to track falsy values at all. Although it's rarely needed, it should be possible to do so if you want to.

For example, CorMisconfigurationForCredentials actually tracks `null` literals for which this barrier is too restrictive (in theory; I'm not aware of any results that would be affected). That happens to be a taint-tracking configuration so it would be affected anyway, but it's a nice example of a query that actually cares about falsy values.

Another way to handle this which we've talked about is to block all nodes that the type inference has inferred to be falsy. So the way to "opt out" might one day be to make it configurable which abstract values you want to track/block. (This is on the JS2020 list and I'm still hoping to get around to it).

In the meantime, I'd suggest making this the default for taint-tracking configurations and expose a nice way to opt-in for data-flow configurations.

**Comment by erik-krogh** [comment]
> In the meantime, I'd suggest making this the default for taint-tracking configurations and expose a nice way to opt-in for data-flow configurations.

:+1:

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 1]
```diff
@@ -1481,8 +1481,11 @@ private class AdditionalBarrierGuardCall extends AdditionalBarrierGuardNode, Dat
   override predicate appliesTo(Configuration cfg) { f.appliesTo(cfg) }
 }
 
-/** A check of the `if(x)`, which sanitizes `x` in its "else" branch. */
-private class VarAccessBarrierGuard extends AdditionalBarrierGuardNode, DataFlow::Node {
+/**
+  * A check of the `if(x)`, which sanitizes `x` in its "else" branch.
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]:
```suggestion
  * A check of the form `if(x)`, which sanitizes `x` in its "else" branch. 
```

**Comment by erik-krogh** [comment]
Evaluation came back bad, I'll look into it.

**Comment by erik-krogh** [comment]
Seems like the bad evaluation was due to the misoptimization in PrototypePollutionUtility that got fixed in https://github.com/Semmle/ql/pull/2779.  

I'm running another evaluation.

**Comment by erik-krogh** [comment]
The [evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1581487773795) ([part2](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-asger.northeurope.cloudapp.azure.com_1581513586710)) is looking like the performance hit is too great for the benefit we get.  

There is 1 instance of `VarAccessBarrierGuard` for every `VarAccess`, and I'm thinking that might be causing the performance regression. 
I'll try to use `UselessConditional::isExplicitConditional` to limit the number of `VarAccessBarrierGuard` instances.

**Comment by erik-krogh** [comment]
An [evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-asger.northeurope.cloudapp.azure.com_1581842134708) over the weekend looks in terms of performance, except `gecko-dev` seemingly timing out. (I will look at that)

**Comment by asgerf** [comment]
Maybe we should try the same solution I used in PrototypePollutionUtility, in which the guard node for `x` is directly marked as a barrier instead of using a barrier _guard_. The difference is that barrier guards trigger reasoning about dominators, which is redundant in this case where SSA encodes dominance information anyway.

**Comment by erik-krogh** [comment]
> Maybe we should try the same solution I used in PrototypePollutionUtility, in which the guard node for x is directly marked as a barrier instead of using a barrier guard.

I've done this now, and performance seems fine from a first look (I forgot --dpm on that run, and performance was ±10%, so I'm redoing with --dpm).  

Although I don't fancy the `or` i placed in `TaintTracking::Configuration::isBarrier`.

**Comment by erik-krogh** [comment]
[Performance with --dpm looks good](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1582097015303).  
(The `amphtml` performance result is a leftover that I didn't clear).

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 2]
```diff
@@ -1483,16 +1483,15 @@ private class AdditionalBarrierGuardCall extends AdditionalBarrierGuardNode, Dat
 
 /**
   * A check of the form `if(x)`, which sanitizes `x` in its "else" branch.
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]:
```suggestion
  * A guard node for a variable in a negative condition, such as `x` in `if(!x)`.
```

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 3]
```diff
@@ -1483,16 +1483,15 @@ private class AdditionalBarrierGuardCall extends AdditionalBarrierGuardNode, Dat
 
 /**
   * A check of the form `if(x)`, which sanitizes `x` in its "else" branch. 
-  * Can be added to a `isBarrierGuard` in a configuration to add the sanitization. 
+  * Can be added to a `isBarrier` in a configuration to add the sanitization.
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]:
```suggestion
  * Can be added to a `isBarrier` in a data-flow configuration to block flow through such checks.
```

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll [hunk 4]
```diff
@@ -355,6 +355,11 @@ module TaintedPath {
     }
   }
 
+  /**
+   * A check of the form `if(x)`, which sanitizes `x` in its "else" branch.
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll]: ditto

Could you make the class private?

**Pull request merged**